File: java/structures/QueueMin.java
package structures;

import java.util.*;

// https://cp-algorithms.com/data_structures/stack_queue_modification.html
public class QueueMin {
    List<Integer> s1 = new ArrayList<>();
    List<Integer> s2 = new ArrayList<>();
    int min1 = Integer.MAX_VALUE;

    public int min() {
        return Math.min(s2.isEmpty() ? Integer.MAX_VALUE : s2.get(s2.size() - 1), min1);
    }

    public void addLast(int x) {
        s1.add(x);
        min1 = Math.min(min1, x);
    }

    public void removeFirst() {
        while (!s1.isEmpty()) {
            int x = s1.remove(s1.size() - 1);
            s2.add(s2.isEmpty() ? x : Math.min(x, s2.get(s2.size() - 1)));
        }
        min1 = Integer.MAX_VALUE;
        s2.remove(s2.size() - 1);
    }

    // Usage example
    public static void main(String[] args) {
        QueueMin q = new QueueMin();
        q.addLast(2);
        q.addLast(3);
        System.out.println(2 == q.min());
        q.removeFirst();
        System.out.println(3 == q.min());
        q.addLast(1);
        System.out.println(1 == q.min());
    }
}

File: cpp/structures/queue_min.cpp
#include <bits/stdc++.h>

using namespace std;

// https://cp-algorithms.com/data_structures/stack_queue_modification.html

template <class T>
struct queue_min {
    stack<pair<T, T>> s1;
    stack<pair<T, T>> s2;

    T min() {
        return std::min(s1.empty() ? numeric_limits<T>::max() : s1.top().second,
                        s2.empty() ? numeric_limits<T>::max() : s2.top().second);
    }

    void add_last(T x) {
        T min_value = s1.empty() ? x : std::min(x, s1.top().second);
        s1.emplace(x, min_value);
    }

    T remove_first() {
        if (s2.empty()) {
            while (!s1.empty()) {
                T x = s1.top().first;
                s1.pop();
                T min_value = s2.empty() ? x : std::min(x, s2.top().second);
                s2.emplace(x, min_value);
            }
        }
        T x = s2.top().first;
        s2.pop();
        return x;
    }
};

// usage example
int main() {
    queue_min<int> q;
    q.add_last(2);
    q.add_last(3);
    cout << boolalpha;
    cout << (2 == q.min()) << endl;
    q.remove_first();
    cout << (3 == q.min()) << endl;
    q.add_last(1);
    cout << (1 == q.min()) << endl;
}
